grails-app/services/org/pih/warehouse/requisition/RequisitionService.groovy

```diff
@@ -842,6 +842,24 @@ class RequisitionService {
                 requisition.status = RequisitionStatus.VERIFYING
                 requisition.approvalRequired = false
                 requisition.save()
+            } else if (status == RequisitionStatus.APPROVED) {
+                requisition.dateApproved = new Date()
+                requisition.approvedBy = currentUser
+                requisition.status = RequisitionStatus.APPROVED
+
+                Event event = createEvent(EventCode.APPROVED, requisition.origin, new Date(), currentUser)
+                requisition.addToEvents(event)
+                requisition.save()
+                return
+            } else if (status == RequisitionStatus.REJECTED) {
+                requisition.dateRejected = new Date()
+                requisition.rejectedBy = currentUser
+                requisition.status = RequisitionStatus.REJECTED
+
+                Event event = createEvent(EventCode.REJECTED, requisition.origin, new Date(), currentUser)
+                requisition.addToEvents(event)
+                requisition.save()
+                return
```

Comment:
I think this could be refactored. We have almost the same content in all of the statements, hence we could create a helper method to handle that, something like:
```groovy
Requisition transitionRequisitionStatus(Requisition requisition, RequisitionStatus requisitionStatus, EventCode eventCode, User currentUser) {
   requisition.dateApproved = new Date()
   requisition.approvedBy = currentUser
   requisition.status = requisitionStatus

   Event event = createEvent(eventCode, requisition.origin, new Date(), currentUser)
   requisition.addToEvents(event)
   requisition.save()
}
```
and here:
```groovy
if (status == RequisitionStatus.PENDING_APPROVAL) {
    if (requisition.origin.approvalRequired) {
        Event event = createEvent(EventCode.PENDING_APPROVAL, requisition.origin, new Date(), currentUser)
        requisition.addToEvents(event)
        requisition.status = RequisitionStatus.PENDING_APPROVAL
        requisition.approvalRequired = true
        requisition.save()
        return
    }
    Event event = createEvent(EventCode.SUBMITTED, requisition.origin, new Date(), currentUser)
    requisition.addToEvents(event)
    requisition.status = RequisitionStatus.VERIFYING
    requisition.approvalRequired = false
    requisition.save()
    return
} 
if (status == RequisitionStatus.APPROVED) {
   transitionRequisitionStatus(requisition, RequisitionStatus.APPROVED, EventCode.APPROVED, currentUser)
   return
}
if (status == RequisitionStatus.REJECTED) {
   transitionRequisitionStatus(requisition, RequisitionStatus.REJECTED, EventCode.REJECTED, currentUser)
   return
}
```
we could go even further and you could refactor also the statements I've made with one minor change:
```groovy
if (status == RequisitionStatus.PENDING_APPROVAL) {
    if (requisition.origin.approvalRequired) {
       Requisition updatedRequisition = transitionRequisitionStatus(requisition, RequisitionStatus.PENDING_APPROVAL, EventCode.PENDING_APPROVAL, currentUser)
       updatedRequisition.approvalRequired = true
       // requisition.save() <--- notice I've commented that, it shouldn't be necessary anymore, because the event which has not been yet persisted has been already saved in the transition method, hence another save should not be necessary and the approvalRequired flag should be persisted at the end of the session, but please, double check that.
       return
    }
    Requisition updatedRequisition = transitionRequisitionStatus(requisition, RequisitionStatus.VERIFYING, EventCode.SUBMITTED, currentUser)
    requisition.approvalRequired = false
    return
} 
if (status == RequisitionStatus.APPROVED) {
   transitionRequisitionStatus(requisition, RequisitionStatus.APPROVED, EventCode.APPROVED, currentUser)
   return
}
if (status == RequisitionStatus.REJECTED) {
   transitionRequisitionStatus(requisition, RequisitionStatus.REJECTED, EventCode.REJECTED, currentUser)
   return
}
```

Reply:
It would be great to move some common logic into some other method but the reason why I didn't do it at first was that there a slight differences in each of the status changes like in example of a APPROVED we assign `approvedBy` but in case of the REJECETD we must assign `rejectedBy`.
These are small differences so a common method should work but I'll think about if there is a better implementation for all of this

Reply:
then you can just follow the path I've shown with my statements with the `approvalRequired` - so just from this method return the common stuff, and then just operate on it like:
```groovy
Requisition updatedRequisition = transitionRequisitionStatus(...)
requisition.rejectedBy = currentUser
```

Reply:
I would also be happy with a switch statement for the top level status check 

```
switch (status) { 
    case PENDING_APPROVAL: 
        doPendingApprovalStuff()
        return
    case APPROVED: 
        transitionRequisitionStatus(requisition, RequisitionStatus.APPROVED, EventCode.APPROVED, currentUser)
        return
    ...  
}
```

Reply:
> requisition.save() <--- notice I've commented that, it shouldn't be necessary anymore, because the event which has not been yet persisted has been already saved in the transition method, hence another save should not be necessary and the approvalRequired flag should be persisted at the end of the session, but please, double check that.

Not sure if that'll be true on Grails 3.

Reply:
I like the switch case a lot more

Reply:
@jmiranda 

> Not sure if that'll be true on Grails 3.

yes, it will, because it's Hibernate's mechanism, not Grails'